firestore/query.go

```diff
@@ -250,6 +251,159 @@ func (q *Query) processCursorArg(name string, docSnapshotOrFieldValues []interfa
 
 func (q Query) query() *Query { return &q }
 
+func getSafeCursorValue(vProto *pb.Value, q Query) (interface{}, error) {
+	switch v := vProto.ValueType.(type) {
+	case *pb.Value_ReferenceValue:
+		refVal := vProto.GetReferenceValue()
+		if strings.HasPrefix(refVal, q.path) {
+			refVal = refVal[len(q.path)+1:]
+		}
+		return refVal, nil
+	case *pb.Value_IntegerValue:
+		// TODO: this shouldn't be needed, but tests send things as int, this returns
+		// int64, so diff fails
+		return v.IntegerValue, nil //(v.IntegerValue), nil
+	default:
+		i, err := createFromProtoValue(vProto, q.c)
+		if err != nil {
+			q.err = err
+			return q, err
+		}
+		return i, nil
+	}
+}
+
+// FromProto creates a new Query object from a RunQueryRequest. This can be used
+// in combintation with ToProto to serialize Query objects.
+func (q Query) FromProto(pbQuery *pb.RunQueryRequest) (Query, error) {
+	// Ensure we are starting from an empty query, but with this client.
+	q = *q.c.Query()
+
+	// 	err                    error
+	q.err = nil
+
+	pbq := pbQuery.GetStructuredQuery()
+	if from := pbq.GetFrom(); len(from) > 0 {
+		// collectionID           string
+		q.collectionID = from[0].CollectionId
+		// allDescendants indicates whether this query is for all collections
+		// that match the ID under the specified parentPath.
+		q.allDescendants = from[0].AllDescendants
+	}
+
+	// 	path                   string // path to query (collection)
+	// 	parentPath             string // path of the collection's parent (document)
+	parent := pbQuery.GetParent()
+	q.parentPath = parent
+	q.path = parent + "/" + q.collectionID
+
+	// 	startVals, endVals     []interface{}
+	// 	startDoc, endDoc       *DocumentSnapshot
+	// 	startBefore, endBefore bool
+	if startAt := pbq.GetStartAt(); startAt != nil {
+		if startAt.GetBefore() {
+			q.startBefore = true
+		}
+		for _, v := range startAt.GetValues() {
+			c, err := getSafeCursorValue(v, q)
```

Comment:
I am not sure how difficult it would be, but this conversion would not be needed if the Query class stored the cursor as the Proto.

Reply:
The change to back cursors by Protos should actually be very simple and remove `getSafeCursorValue()`. Unlike filters and orderBys, you can store the cursor directly and do not need to wrap it in an array.

Reply:
I made these changes and I like them, unfortunately the go library allows orderby after startat, which means we need to have a doc snapshot available to retrieve fields from. As removing this would be breaking I am not going to be simplifying this now.

Reply:
https://github.com/googleapis/google-cloud-go/issues/4448 to track that breaking change though. I'd like to do this when we need to do a break so we would enable this simplification in the future.